The app crashed when I tried to convert a password protected PDF file to Image

#### Expected behavior
The users should be able to convert a password protected PDF file to Image without crashing the app.

#### Actual behavior
When I tried to convert a password protected PDF file to Image, sometimes it crashed immediately/ after a few attempts.

#### How to reproduce
1. Open the app.
2. Then go to "PDF to Images".
3. Swipe up the view files menu.
4. Choose and click on a password protected PDF file. (If you don't have any password protected file then download from somewhere or create a new one from "Add password" menu).
5. Then click on "CREATE IMAGES" and notice the crash. (Sometimes it crashes after a few attempts.)

* App version: 8.6.5
* Operating system: Android 6.0.1

#### Recording Of The Bug
![imagetopdf](https://media.giphy.com/media/XfKACzxiOhmm8Y0iSz/giphy.gif)

#### LogCat
```
04-22 12:15:35.743 22730 22788 E AndroidRuntime: FATAL EXCEPTION: AsyncTask #4

04-22 12:15:35.743 22730 22788 E AndroidRuntime: Process: swati4star.createpdf, PID: 22730

04-22 12:15:35.743 22730 22788 E AndroidRuntime: java.lang.RuntimeException: An error occurred while executing doInBackground()

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at android.os.AsyncTask$3.done(AsyncTask.java:309)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at java.util.concurrent.FutureTask.setException(FutureTask.java:223)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at java.util.concurrent.FutureTask.run(FutureTask.java:242)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:234)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at java.lang.Thread.run(Thread.java:818)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: Caused by: java.lang.IllegalStateException: cannot load page

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at android.graphics.pdf.PdfRenderer.nativeOpenPageAndGetSize(Native Method)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at android.graphics.pdf.PdfRenderer.access$300(PdfRenderer.java:101)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at android.graphics.pdf.PdfRenderer$Page.<init>(PdfRenderer.java:276)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at android.graphics.pdf.PdfRenderer$Page.<init>(PdfRenderer.java:254)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at android.graphics.pdf.PdfRenderer.openPage(PdfRenderer.java:203)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at swati4star.createpdf.util.PdfToImages.doInBackground(PdfToImages.java:62)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at swati4star.createpdf.util.PdfToImages.doInBackground(PdfToImages.java:22)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at android.os.AsyncTask$2.call(AsyncTask.java:295)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)

04-22 12:15:35.743 22730 22788 E AndroidRuntime: 	... 4 more
```

--- a/app/src/main/java/swati4star/createpdf/fragment/PdfToImageFragment.java
+++ b/app/src/main/java/swati4star/createpdf/fragment/PdfToImageFragment.java
@@ -10,6 +10,9 @@ import android.support.annotation.NonNull;
 import android.support.design.widget.BottomSheetBehavior;
 import android.support.v4.app.Fragment;
 import android.support.v7.widget.RecyclerView;
+import android.text.Editable;
+import android.text.TextWatcher;
+import android.util.Log;
 import android.view.LayoutInflater;
 import android.view.View;
 import android.view.ViewGroup;
@@ -41,8 +44,11 @@ import swati4star.createpdf.util.CommonCodeUtils;
 import swati4star.createpdf.util.ExtractImages;
 import swati4star.createpdf.util.FileUtils;
 import swati4star.createpdf.util.MorphButtonUtility;
+import swati4star.createpdf.util.PDFEncryptionUtility;
+import swati4star.createpdf.util.PDFUtils;
 import swati4star.createpdf.util.PdfToImages;
 import swati4star.createpdf.util.RealPathUtil;
+import swati4star.createpdf.util.StringUtils;
 
 import static android.app.Activity.RESULT_OK;
 import static swati4star.createpdf.util.CommonCodeUtils.checkSheetBehaviourUtil;
@@ -51,10 +57,12 @@ import static swati4star.createpdf.util.CommonCodeUtils.populateUtil;
 import static swati4star.createpdf.util.Constants.BUNDLE_DATA;
 import static swati4star.createpdf.util.Constants.PDF_TO_IMAGES;
 import static swati4star.createpdf.util.DialogUtils.createAnimationDialog;
+import static swati4star.createpdf.util.DialogUtils.createOverwriteDialog;
 import static swati4star.createpdf.util.StringUtils.showSnackbar;
 
 public class PdfToImageFragment extends Fragment implements BottomSheetPopulate, MergeFilesAdapter.OnClickListener,
-        ExtractImagesListener, ExtractImagesAdapter.OnFileItemClickedListener, OnBackPressedInterface {
+        ExtractImagesListener, ExtractImagesAdapter.OnFileItemClickedListener,
+        OnBackPressedInterface {
 
     private static final int INTENT_REQUEST_PICKFILE_CODE = 10;
     private Activity mActivity;
@@ -67,6 +75,9 @@ public class PdfToImageFragment extends Fragment implements BottomSheetPopulate,
     private ArrayList<String> mOutputFilePaths;
     private MaterialDialog mMaterialDialog;
     private String mOperation;
+    private Context mContext;
+    private PDFUtils mPDFUtils;
+    private String[] mInputPassword;
 
     @BindView(R.id.lottie_progress)
     LottieAnimationView mLottieProgress;
@@ -177,14 +188,57 @@ public class PdfToImageFragment extends Fragment implements BottomSheetPopulate,
     }
 
     /**
-     * invokes generation of images for pdf pages in the background
+     * invokes generation of images for pdf pages in the background by checking
+     * for encryption first.
      */
     @OnClick(R.id.createImages)
     public void parse() {
-        if (mOperation.equals(PDF_TO_IMAGES))
-            new PdfToImages(mPath, mUri, this).execute();
-        else
-            new ExtractImages(mPath, this).execute();
+        if (mPDFUtils.isPDFEncrypted(mPath)) {
+            mInputPassword = new String[1];
+            new MaterialDialog.Builder(mActivity)
+                    .title(R.string.enter_password)
+                    .content(R.string.decrypt_message)
+                    .input(null, null, (dialog, input) -> {
+                        if (StringUtils.isEmpty(input)) {
+                            showSnackbar(mActivity, R.string.snackbar_name_not_blank);
+                        } else {
+                            final String inputName = input.toString();
+                            if (inputName != null) {
+                                mInputPassword[0] = inputName;
+                                pdfToImage(mInputPassword);
+                            } else {
+                                MaterialDialog.Builder builder = createOverwriteDialog(mActivity);
+                                builder.onPositive((dialog12, which) -> {
+                                    mInputPassword[0] = inputName;
+                                    pdfToImage(mInputPassword);
+                                })
+                                        .onNegative((dialog1, which) -> parse())
+                                        .show();
+                            }
+                        }
+                    })
+                    .show();
+        }
+    }
+
+    /**
+     * Thia method handles the call to the Async process of conversion
+     * from PDF to Image.
+     * @param mInputPassword is the password if the file is encrypted.
+     */
+    private void pdfToImage(String[] mInputPassword) {
+        if (mOperation.equals(PDF_TO_IMAGES)) {
+            if (mInputPassword != null) {
+                new PdfToImages(mContext, mInputPassword,
+                        mPath, mUri, this)
+                        .execute();
+            } else {
+                new PdfToImages(mContext, mPath, mUri, PdfToImageFragment.this)
+                        .execute();
+            }
+        } else
+            new ExtractImages(mContext, mPath, PdfToImageFragment.this)
+                    .execute();
     }
 
     @Override
@@ -194,6 +248,8 @@ public class PdfToImageFragment extends Fragment implements BottomSheetPopulate,
         mMorphButtonUtility = new MorphButtonUtility(mActivity);
         mFileUtils = new FileUtils(mActivity);
         mBottomSheetUtils = new BottomSheetUtils(mActivity);
+        mContext = context;
+        mPDFUtils = new PDFUtils(mActivity);
     }
 
     /**

--- a/app/src/main/java/swati4star/createpdf/fragment/PdfToImageFragment.java
+++ b/app/src/main/java/swati4star/createpdf/fragment/PdfToImageFragment.java
@@ -10,9 +10,6 @@ import android.support.annotation.NonNull;
 import android.support.design.widget.BottomSheetBehavior;
 import android.support.v4.app.Fragment;
 import android.support.v7.widget.RecyclerView;
-import android.text.Editable;
-import android.text.TextWatcher;
-import android.util.Log;
 import android.view.LayoutInflater;
 import android.view.View;
 import android.view.ViewGroup;
@@ -44,7 +41,6 @@ import swati4star.createpdf.util.CommonCodeUtils;
 import swati4star.createpdf.util.ExtractImages;
 import swati4star.createpdf.util.FileUtils;
 import swati4star.createpdf.util.MorphButtonUtility;
-import swati4star.createpdf.util.PDFEncryptionUtility;
 import swati4star.createpdf.util.PDFUtils;
 import swati4star.createpdf.util.PdfToImages;
 import swati4star.createpdf.util.RealPathUtil;
@@ -61,8 +57,7 @@ import static swati4star.createpdf.util.DialogUtils.createOverwriteDialog;
 import static swati4star.createpdf.util.StringUtils.showSnackbar;
 
 public class PdfToImageFragment extends Fragment implements BottomSheetPopulate, MergeFilesAdapter.OnClickListener,
-        ExtractImagesListener, ExtractImagesAdapter.OnFileItemClickedListener,
-        OnBackPressedInterface {
+        ExtractImagesListener, ExtractImagesAdapter.OnFileItemClickedListener, OnBackPressedInterface {
 
     private static final int INTENT_REQUEST_PICKFILE_CODE = 10;
     private Activity mActivity;
@@ -218,12 +213,15 @@ public class PdfToImageFragment extends Fragment implements BottomSheetPopulate,
                         }
                     })
                     .show();
+        } else {
+            pdfToImage(mInputPassword);
         }
     }
 
     /**
      * Thia method handles the call to the Async process of conversion
      * from PDF to Image.
+     *
      * @param mInputPassword is the password if the file is encrypted.
      */
     private void pdfToImage(String[] mInputPassword) {
@@ -233,12 +231,10 @@ public class PdfToImageFragment extends Fragment implements BottomSheetPopulate,
                         mPath, mUri, this)
                         .execute();
             } else {
-                new PdfToImages(mContext, mPath, mUri, PdfToImageFragment.this)
-                        .execute();
+                new PdfToImages(mContext, mPath, mUri, this).execute();
             }
         } else
-            new ExtractImages(mContext, mPath, PdfToImageFragment.this)
-                    .execute();
+            new ExtractImages(mContext, mPath, this).execute();
     }
 
     @Override

--- a/app/src/main/java/swati4star/createpdf/fragment/PdfToImageFragment.java
+++ b/app/src/main/java/swati4star/createpdf/fragment/PdfToImageFragment.java
@@ -10,6 +10,7 @@ import android.support.annotation.NonNull;
 import android.support.design.widget.BottomSheetBehavior;
 import android.support.v4.app.Fragment;
 import android.support.v7.widget.RecyclerView;
+import android.text.InputType;
 import android.view.LayoutInflater;
 import android.view.View;
 import android.view.ViewGroup;
@@ -193,6 +194,7 @@ public class PdfToImageFragment extends Fragment implements BottomSheetPopulate,
             new MaterialDialog.Builder(mActivity)
                     .title(R.string.enter_password)
                     .content(R.string.decrypt_message)
+                    .inputType(InputType.TYPE_TEXT_VARIATION_PASSWORD)
                     .input(null, null, (dialog, input) -> {
                         if (StringUtils.isEmpty(input)) {
                             showSnackbar(mActivity, R.string.snackbar_name_not_blank);
@@ -201,14 +203,6 @@ public class PdfToImageFragment extends Fragment implements BottomSheetPopulate,
                             if (inputName != null) {
                                 mInputPassword[0] = inputName;
                                 pdfToImage(mInputPassword);
-                            } else {
-                                MaterialDialog.Builder builder = createOverwriteDialog(mActivity);
-                                builder.onPositive((dialog12, which) -> {
-                                    mInputPassword[0] = inputName;
-                                    pdfToImage(mInputPassword);
-                                })
-                                        .onNegative((dialog1, which) -> parse())
-                                        .show();
                             }
                         }
                     })
@@ -222,17 +216,12 @@ public class PdfToImageFragment extends Fragment implements BottomSheetPopulate,
      * Thia method handles the call to the Async process of conversion
      * from PDF to Image.
      *
-     * @param mInputPassword is the password if the file is encrypted.
+     * @param mInputPassword - the password if the file is encrypted.
      */
     private void pdfToImage(String[] mInputPassword) {
         if (mOperation.equals(PDF_TO_IMAGES)) {
-            if (mInputPassword != null) {
-                new PdfToImages(mContext, mInputPassword,
-                        mPath, mUri, this)
-                        .execute();
-            } else {
-                new PdfToImages(mContext, mPath, mUri, this).execute();
-            }
+            new PdfToImages(mContext, mInputPassword, mPath, mUri, this)
+                    .execute();
         } else
             new ExtractImages(mContext, mPath, this).execute();
     }

--- a/app/src/main/java/swati4star/createpdf/fragment/PdfToImageFragment.java
+++ b/app/src/main/java/swati4star/createpdf/fragment/PdfToImageFragment.java
@@ -193,7 +193,7 @@ public class PdfToImageFragment extends Fragment implements BottomSheetPopulate,
             mInputPassword = new String[1];
             new MaterialDialog.Builder(mActivity)
                     .title(R.string.enter_password)
-                    .content(R.string.decrypt_message)
+                    .content(R.string.decrypt_protected_file)
                     .inputType(InputType.TYPE_TEXT_VARIATION_PASSWORD)
                     .input(null, null, (dialog, input) -> {
                         if (StringUtils.isEmpty(input)) {
@@ -223,7 +223,7 @@ public class PdfToImageFragment extends Fragment implements BottomSheetPopulate,
             new PdfToImages(mContext, mInputPassword, mPath, mUri, this)
                     .execute();
         } else
-            new ExtractImages(mContext, mPath, this).execute();
+            new ExtractImages(mPath, this).execute();
     }
 
     @Override

--- a/app/src/main/java/swati4star/createpdf/util/PDFEncryptionUtility.java
+++ b/app/src/main/java/swati4star/createpdf/util/PDFEncryptionUtility.java
@@ -63,8 +63,7 @@ public class PDFEncryptionUtility {
         mPasswordInput.addTextChangedListener(
                 new TextWatcher() {
                     @Override
-                    public void beforeTextChanged(CharSequence s, int start, int count, int after) {
-                    }
+                    public void beforeTextChanged(CharSequence s, int start, int count, int after) { }
 
                     @Override
                     public void onTextChanged(CharSequence s, int start, int before, int count) {
@@ -194,9 +193,9 @@ public class PDFEncryptionUtility {
 
     /**
      * This function removes the password for encrypted files.
-     * @param file is the path of the actual encrypted file.
-     * @param dataSetChanged is the DataSetChanged object.
-     * @param inputPassword is the password of the encrypted file.
+     * @param file - the path of the actual encrypted file.
+     * @param dataSetChanged - the DataSetChanged object.
+     * @param inputPassword - the password of the encrypted file.
      * @return
      */
     public String removeDefPasswordForImages(final String file,
@@ -216,9 +215,6 @@ public class PDFEncryptionUtility {
                 PdfStamper stamper = new PdfStamper(reader, new FileOutputStream(finalOutputFile));
                 stamper.close();
                 reader.close();
-                if (dataSetChanged != null)
-                    dataSetChanged.updateDataset();
-                new DatabaseHelper(mContext).insertRecord(finalOutputFile, mContext.getString(R.string.decrypted));
                 String filepath = finalOutputFile;
                 return filepath;
             }

--- a/app/src/main/java/swati4star/createpdf/util/PDFEncryptionUtility.java
+++ b/app/src/main/java/swati4star/createpdf/util/PDFEncryptionUtility.java
@@ -194,12 +194,10 @@ public class PDFEncryptionUtility {
     /**
      * This function removes the password for encrypted files.
      * @param file - the path of the actual encrypted file.
-     * @param dataSetChanged - the DataSetChanged object.
      * @param inputPassword - the password of the encrypted file.
      * @return
      */
     public String removeDefPasswordForImages(final String file,
-                                             final DataSetChanged dataSetChanged,
                                              final String[] inputPassword) {
         String finalOutputFile;
         try {
@@ -215,8 +213,7 @@ public class PDFEncryptionUtility {
                 PdfStamper stamper = new PdfStamper(reader, new FileOutputStream(finalOutputFile));
                 stamper.close();
                 reader.close();
-                String filepath = finalOutputFile;
-                return filepath;
+                return finalOutputFile;
             }
         } catch (DocumentException | IOException e) {
             e.printStackTrace();

--- a/app/src/main/java/swati4star/createpdf/util/PDFEncryptionUtility.java
+++ b/app/src/main/java/swati4star/createpdf/util/PDFEncryptionUtility.java
@@ -24,7 +24,6 @@ import java.util.Objects;
 import swati4star.createpdf.R;
 import swati4star.createpdf.database.DatabaseHelper;
 import swati4star.createpdf.interfaces.DataSetChanged;
-import swati4star.createpdf.interfaces.OnPDFDecryptInterface;
 
 import static swati4star.createpdf.util.Constants.MASTER_PWD_STRING;
 import static swati4star.createpdf.util.Constants.appName;

--- a/app/src/main/java/swati4star/createpdf/util/PDFEncryptionUtility.java
+++ b/app/src/main/java/swati4star/createpdf/util/PDFEncryptionUtility.java
@@ -24,6 +24,7 @@ import java.util.Objects;
 import swati4star.createpdf.R;
 import swati4star.createpdf.database.DatabaseHelper;
 import swati4star.createpdf.interfaces.DataSetChanged;
+import swati4star.createpdf.interfaces.OnPDFDecryptInterface;
 
 import static swati4star.createpdf.util.Constants.MASTER_PWD_STRING;
 import static swati4star.createpdf.util.Constants.appName;
@@ -63,7 +64,8 @@ public class PDFEncryptionUtility {
         mPasswordInput.addTextChangedListener(
                 new TextWatcher() {
                     @Override
-                    public void beforeTextChanged(CharSequence s, int start, int count, int after) { }
+                    public void beforeTextChanged(CharSequence s, int start, int count, int after) {
+                    }
 
                     @Override
                     public void onTextChanged(CharSequence s, int start, int before, int count) {
@@ -189,6 +191,42 @@ public class PDFEncryptionUtility {
             }
             mDialog.dismiss();
         });
+    }
+
+    /**
+     * This function removes the password for encrypted files.
+     * @param file is the path of the actual encrypted file.
+     * @param dataSetChanged is the DataSetChanged object.
+     * @param inputPassword is the password of the encrypted file.
+     * @return
+     */
+    public String removeDefPasswordForImages(final String file,
+                                             final DataSetChanged dataSetChanged,
+                                             final String[] inputPassword) {
+        String finalOutputFile;
+        try {
+            String masterpwd = mSharedPrefs.getString(MASTER_PWD_STRING, appName);
+            PdfReader reader = new PdfReader(file, masterpwd.getBytes());
+            byte[] password;
+            finalOutputFile = mFileUtils.getUniqueFileName
+                    (file.replace(mContext.getResources().getString(R.string.pdf_ext),
+                            mContext.getString(R.string.decrypted_file)));
+            password = reader.computeUserPassword();
+            byte[] input = inputPassword[0].getBytes();
+            if (Arrays.equals(input, password)) {
+                PdfStamper stamper = new PdfStamper(reader, new FileOutputStream(finalOutputFile));
+                stamper.close();
+                reader.close();
+                if (dataSetChanged != null)
+                    dataSetChanged.updateDataset();
+                new DatabaseHelper(mContext).insertRecord(finalOutputFile, mContext.getString(R.string.decrypted));
+                String filepath = finalOutputFile;
+                return filepath;
+            }
+        } catch (DocumentException | IOException e) {
+            e.printStackTrace();
+        }
+        return null;
     }
 
     private boolean removePasswordUsingDefMasterPAssword(final String file,

--- a/app/src/main/java/swati4star/createpdf/util/PdfToImages.java
+++ b/app/src/main/java/swati4star/createpdf/util/PdfToImages.java
@@ -33,15 +33,6 @@ public class PdfToImages extends AsyncTask<Void, Void, Void> {
     private Context mContext;
     private String mDecryptedPath;
 
-    public PdfToImages(Context context, String mPath, Uri mUri,
-                       ExtractImagesListener mExtractImagesListener) {
-        this.mPath = mPath;
-        this.mUri = mUri;
-        this.mExtractImagesListener = mExtractImagesListener;
-        mOutputFilePaths = new ArrayList<>();
-        this.mContext = context;
-    }
-
     public PdfToImages(Context context, String[] password, String mPath, Uri mUri,
                        ExtractImagesListener mExtractImagesListener) {
         this.mPath = mPath;
@@ -123,5 +114,7 @@ public class PdfToImages extends AsyncTask<Void, Void, Void> {
     protected void onPostExecute(Void aVoid) {
         super.onPostExecute(aVoid);
         mExtractImagesListener.updateView(mImagesCount, mOutputFilePaths);
+        if (mDecryptedPath != null)
+            new File(mDecryptedPath).delete();
     }
 }

--- a/app/src/main/java/swati4star/createpdf/util/PdfToImages.java
+++ b/app/src/main/java/swati4star/createpdf/util/PdfToImages.java
@@ -1,5 +1,7 @@
 package swati4star.createpdf.util;
 
+import android.app.Activity;
+import android.content.Context;
 import android.graphics.Bitmap;
 import android.graphics.Canvas;
 import android.graphics.Color;
@@ -26,35 +28,61 @@ public class PdfToImages extends AsyncTask<Void, Void, Void> {
     private ExtractImagesListener mExtractImagesListener;
     private int mImagesCount = 0;
     private ArrayList<String> mOutputFilePaths;
+    private String[] mPassword;
+    private PDFEncryptionUtility mPDFEncryptionUtility;
+    private Context mContext;
+    private String mDecryptedPath;
 
-    public PdfToImages(String mPath, Uri mUri, ExtractImagesListener mExtractImagesListener) {
+    public PdfToImages(Context context, String mPath, Uri mUri,
+                       ExtractImagesListener mExtractImagesListener) {
         this.mPath = mPath;
         this.mUri = mUri;
         this.mExtractImagesListener = mExtractImagesListener;
         mOutputFilePaths = new ArrayList<>();
+        this.mContext = context;
+    }
+
+    public PdfToImages(Context context, String[] password, String mPath, Uri mUri,
+                       ExtractImagesListener mExtractImagesListener) {
+        this.mPath = mPath;
+        this.mUri = mUri;
+        this.mExtractImagesListener = mExtractImagesListener;
+        mOutputFilePaths = new ArrayList<>();
+        this.mPassword = password;
+        this.mContext = context;
     }
 
     @Override
     protected void onPreExecute() {
         super.onPreExecute();
+        mPDFEncryptionUtility = new PDFEncryptionUtility((Activity) mContext);
         mExtractImagesListener.extractionStarted();
     }
 
     @Override
     protected Void doInBackground(Void... voids) {
+        if (mPassword != null) {
+            mDecryptedPath = mPDFEncryptionUtility.removeDefPasswordForImages(mPath,
+                    null, mPassword);
+        }
         mOutputFilePaths = new ArrayList<>();
         mImagesCount = 0;
 
         // Render pdf pages as bitmap
         ParcelFileDescriptor fileDescriptor = null;
         try {
-            if (mUri != null)
-                // resolve pdf file path based on uri
-                fileDescriptor = ((PdfToImageFragment) mExtractImagesListener).getContext()
-                        .getContentResolver().openFileDescriptor(mUri, "r");
-            else if (mPath != null)
-                // resolve pdf file path based on relative path
-                fileDescriptor = ParcelFileDescriptor.open(new File(mPath), MODE_READ_ONLY);
+            if (mDecryptedPath != null)
+                fileDescriptor = ParcelFileDescriptor.open(new File(mDecryptedPath), MODE_READ_ONLY);
+            else {
+                if (mUri != null) {
+                    // resolve pdf file path based on uri
+                    fileDescriptor = ((PdfToImageFragment) mExtractImagesListener).getContext()
+                            .getContentResolver().openFileDescriptor(mUri, "r");
+                } else if (mPath != null) {
+                    // resolve pdf file path based on relative path
+                    fileDescriptor = ParcelFileDescriptor.open(new File(mPath), MODE_READ_ONLY);
+                }
+            }
             if (fileDescriptor != null) {
                 PdfRenderer renderer = new PdfRenderer(fileDescriptor);
                 final int pageCount = renderer.getPageCount();
@@ -88,8 +116,6 @@ public class PdfToImages extends AsyncTask<Void, Void, Void> {
         } catch (IOException | SecurityException e) {
             e.printStackTrace();
         }
-
-
         return null;
     }
 

--- a/app/src/main/java/swati4star/createpdf/util/PdfToImages.java
+++ b/app/src/main/java/swati4star/createpdf/util/PdfToImages.java
@@ -53,8 +53,7 @@ public class PdfToImages extends AsyncTask<Void, Void, Void> {
     @Override
     protected Void doInBackground(Void... voids) {
         if (mPassword != null) {
-            mDecryptedPath = mPDFEncryptionUtility.removeDefPasswordForImages(mPath,
-                    null, mPassword);
+            mDecryptedPath = mPDFEncryptionUtility.removeDefPasswordForImages(mPath, mPassword);
         }
         mOutputFilePaths = new ArrayList<>();
         mImagesCount = 0;

--- a/app/src/main/res/values-es/strings.xml
+++ b/app/src/main/res/values-es/strings.xml
@@ -241,6 +241,7 @@
     <string name="extract_text">Extract Text</string>
     <string name="snackbar_txtExtracted">Text extracted!</string>
     <string name="snackbar_pdfselected">PDF file selected!</string>
+    <string name="decrypt_protected_file">Selected pdf is password protected. Enter the password to continue...</string>
     <string name="creating_txt">Creating Text File</string>
     <string name="filter_brush">Pincel</string>
     <string name="filter_saved">Imagen guardada</string>

--- a/app/src/main/res/values-fr/strings.xml
+++ b/app/src/main/res/values-fr/strings.xml
@@ -405,6 +405,7 @@
     <string name="pdf_selected">PDF file selected: </string>
     <string name="snackbar_txtExtracted">Text extracted!</string>
     <string name="snackbar_pdfselected">PDF file selected!</string>
+    <string name="decrypt_protected_file">Selected pdf is password protected. Enter the password to continue...</string>
     <string name="creating_txt">Creating Text File</string>
 
     <string name="font_color">Font Color</string>

--- a/app/src/main/res/values-ja/strings.xml
+++ b/app/src/main/res/values-ja/strings.xml
@@ -403,6 +403,7 @@
     <string name="pdf_selected">PDF file selected: </string>
     <string name="snackbar_txtExtracted">Text extracted!</string>
     <string name="snackbar_pdfselected">PDF file selected!</string>
+    <string name="decrypt_protected_file">Selected pdf is password protected. Enter the password to continue...</string>
     <string name="creating_txt">Creating Text File</string>
 
     <string name="font_color">Font Color</string>

--- a/app/src/main/res/values-ru/strings.xml
+++ b/app/src/main/res/values-ru/strings.xml
@@ -446,6 +446,7 @@
     <string name="pdf_selected">PDF file selected: </string>
     <string name="snackbar_txtExtracted">Text extracted!</string>
     <string name="snackbar_pdfselected">PDF file selected!</string>
+    <string name="decrypt_protected_file">Selected pdf is password protected. Enter the password to continue...</string>
     <string name="creating_txt">Creating Text File</string>
 
     <string name="font_color">Font Color</string>

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -200,6 +200,7 @@
     <string name="set_as_default">Set as default</string>
     <string name="decrypted_file">_unencrypted.pdf</string>
     <string name="decrypt_message">Enter the password to continue</string>
+    <string name="decrypt_protected_file">Selected pdf is password protected. Enter the password to continue...</string>
     <string name="not_encrypted">PDF is not Encrypted</string>
     <string name="save_current">Save</string>
     <string name="save">Save</string>

